File: local_data_api/secret_manager.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Optional, Dict
from hashlib import sha1

from local_data_api.exceptions import BadRequestException


@dataclass
class Secret:
    user_name: Optional[str] = None
    password: Optional[str] = None


SECRETS: Dict[str, Secret] = {}


def create_secret_arn(region_name: str = 'us-east-1', account: str = '123456789012') -> str:
    return f'arn:aws:secretsmanager:{region_name}:{account}:secret:local-data-api{sha1().hexdigest()}'


def register_secret(user_name: Optional[str], password: Optional[str], secret_arn: Optional[str]) -> str:
    if not secret_arn:
        secret_arn = create_secret_arn()

    SECRETS[secret_arn] = Secret(user_name, password)

    return secret_arn


def get_secret(secret_arn: str) -> Secret:
    if secret_arn in SECRETS:
        return SECRETS[secret_arn]
    raise BadRequestException(f'Error fetching secret {secret_arn} : Secrets Manager can’t find the specified '
                              f'secret. (Service: AWSSecretsManager; Status Code: 400; Error Code: '
                              f'ResourceNotFoundException; Request ID:  00000000-1111-2222-3333-44444444444)')

File: kotlin/local-data-api/src/SecretManager.kt
package com.koxudaxi.local_data_api

import java.security.MessageDigest
import java.util.*

class SecretManager {
    private fun createSecretArn(
        regionName: String = "us-east-1", account: String = "123456789012"
    ): String {
        val hash = MessageDigest.getInstance("SHA-1").digest()
        return "arn:aws:secretsmanager:${regionName}:${account}:secret:local-data-api${hash}"
    }


    fun registerSecret(
        userName: String?,
        password: String?,
        secretArn: String?,
    ): String {
        return (secretArn ?: createSecretArn()).apply {
            secrets[this] = Secret(userName, password)

        }
    }

    companion object {

        fun getSecret(secretArn: String): Secret {
            return secrets[secretArn] ?: throw BadRequestException(
                "Error fetching secret {secret_arn} : Secrets Manager can’t find the specified secret." +
                        " (Service: AWSSecretsManager; Status Code: 400; Error Code: " +
                        "ResourceNotFoundException; Request ID:  00000000-1111-2222-3333-44444444444)"
            )
        }
        private val secrets: MutableMap<String, Secret> = Collections.synchronizedMap(HashMap())
    }

    data class Secret(
        val userName: String?,
        val password: String?,
    )
}

